### PR TITLE
Fix interpolation speed regression

### LINKED ISSUE #1358: Performance regression saving XDMF functions
Hello, I noticed a performance regression in my code when I save a function to an XDMF file. See the following excerpts from my code's output.

Using revision `2d64f7abba00c4a36a34da5d52bed157e83690e2` in my branch `mleoni/overlappingBCs`, which incorporates `master` before `FunctionSpace::dim` was removed, I had
```2021-01-24 11:11:10.235 (  21.122s) [main            ]           XDMFFile.cpp:50    INFO| Opened HDF5 file with id "72057594037927938"
2021-01-24 11:11:10.235 (  21.123s) [main            ]          xdmf_mesh.cpp:197   INFO| Adding mesh to node "/Xdmf/Domain"
2021-01-24 11:11:10.236 (  21.123s) [main            ]          xdmf_mesh.cpp:24    INFO| Adding topology data to node "/Xdmf/Domain/Grid"
2021-01-24 11:11:10.324 (  21.211s) [main            ]          xdmf_mesh.cpp:146   INFO| Adding geometry data to node "/Xdmf/Domain/Grid"
2021-01-24 11:11:10.856 (  21.743s) [main            ]           XDMFFile.cpp:50    INFO| Opened HDF5 file with id "72057594037927939"
2021-01-24 11:11:10.856 (  21.744s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-24 11:11:12.567 (  23.455s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-24 11:11:14.228 (  25.115s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-24 11:11:15.845 (  26.732s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-24 11:11:17.461 (  28.348s) [main            ] NavierStokesSolver.cpp:278   INFO| Solving momentum
```
Notice how the four "Adding function" steps take around 7 seconds.

Using revision `2f8fe84e2b9aca87490c17a751f7a6ccb55344e9` from the same branch, that now tracks the most recent `master`, I get
```2021-01-24 11:27:03.347 (  24.614s) [main            ]           XDMFFile.cpp:50    INFO| Opened HDF5 file with id "72057594037927938"
2021-01-24 11:27:03.348 (  24.615s) [main            ]          xdmf_mesh.cpp:197   INFO| Adding mesh to node "/Xdmf/Domain"
2021-01-24 11:27:03.348 (  24.615s) [main            ]          xdmf_mesh.cpp:25    INFO| Adding topology data to node "/Xdmf/Domain/Grid"
2021-01-24 11:27:03.428 (  24.695s) [main            ]          xdmf_mesh.cpp:146   INFO| Adding geometry data to node "/Xdmf/Domain/Grid"
2021-01-24 11:27:04.060 (  25.327s) [main            ]           XDMFFile.cpp:50    INFO| Opened HDF5 file with id "72057594037927939"
2021-01-24 11:27:04.060 (  25.327s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-24 11:27:25.043 (  46.310s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-24 11:27:43.205 (  64.472s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-24 11:28:01.160 (  82.427s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-24 11:28:19.354 ( 100.621s) [main            ] NavierStokesSolver.cpp:286   INFO| Solving momentum
```
where now the same "Adding function" steps take around ten times longer.

Both tests were run in parallel on 4 processors on my laptop, code compiled in Release mode.
When Eigen was used, I remember experiencing that XDMF function output was very slow when compiled in Debug mode, but much faster in Release mode; furthermore, stopping the Debug version with a debugger while executing that function showed a lot of Eigen-related overhead, suggesting that compile-time optimisation was responsible for the performance gain, so maybe this performance drop might have something to do with the recent switch between Eigen and STL vectors.

### PR DESCRIPTION
**Pull request by jorgensd**
Fixes #1358 and another potential issue in FiniteElement only getting the interpolation matrix from basix once
Some additional bug-fixes in interpolation, and proper error handling for mixed elements or invalid lambda functions.

### CODE REVIEW

**Comment by garth-wells** [comment]
Does this change recover all of the performance?

**Comment by jorgensd** [comment]
> Does this change recover all of the performance?

Average runtime over 4 processors on current branch:
```
~Write function                                                |     1  1.518422  1.518422
```
If @massimiliano-leoni could verify that his runtime has gone down to normal levels as well with this fix, that would be great.

**Comment by massimiliano-leoni** [comment]
Hello, I just tried your branch, this is the result:
```2021-01-26 11:06:10.967 (  24.586s) [main            ]           XDMFFile.cpp:50    INFO| Opened HDF5 file with id "72057594037927938"
2021-01-26 11:06:10.967 (  24.586s) [main            ]          xdmf_mesh.cpp:197   INFO| Adding mesh to node "/Xdmf/Domain"
2021-01-26 11:06:10.967 (  24.587s) [main            ]          xdmf_mesh.cpp:25    INFO| Adding topology data to node "/Xdmf/Domain/Grid"
2021-01-26 11:06:11.046 (  24.665s) [main            ]          xdmf_mesh.cpp:146   INFO| Adding geometry data to node "/Xdmf/Domain/Grid"
2021-01-26 11:06:11.477 (  25.096s) [main            ]           XDMFFile.cpp:50    INFO| Opened HDF5 file with id "72057594037927939"
2021-01-26 11:06:11.477 (  25.097s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-26 11:06:16.917 (  30.537s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-26 11:06:22.092 (  35.712s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-26 11:06:27.299 (  40.919s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-26 11:06:32.477 (  46.096s) [main            ] NavierStokesSolver.cpp:286   INFO| Solving momentum
```
21 seconds in total, so better than the 75 seconds of the current version, but still not back to the 6.5 seconds of the old version.

**Comment by mscroggs** [comment]
I've just moved `eval` outside a cell loop as this hugely reduces how many times `basix::tabulate` will be called

File: cpp/dolfinx/fem/Function.h [hunk 1]
```diff
@@ -397,29 +392,22 @@ class Function
...
     // Interpolate point values on each cell (using last computed value if
     // not continuous, e.g. discontinuous Galerkin methods)
-    Eigen::Array<double, Eigen::Dynamic, 3, Eigen::RowMajor> x(num_dofs_g, 3);
-    Eigen::Array<T, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> values(
-        num_dofs_g, value_size_loc);
     auto map = mesh->topology().index_map(tdim);
     assert(map);
     const std::int32_t num_cells = map->size_local() + map->num_ghosts();
+
+    Eigen::Array<int, Eigen::Dynamic, 1> cells(x_g.rows());
```
Comment (garth-wells) [inline cpp/dolfinx/fem/Function.h:399]: Use `std::vector` here. Especially for 1D arrays used mainly as just data structures we want to move away from Eigen.

File: cpp/dolfinx/fem/Function.h [hunk 2]
```diff
@@ -397,29 +392,22 @@ class Function
...
-      cells = c;
-      eval(x, cells, values);
-
-      // Copy values to array of point values
-      for (int i = 0; i < values.rows(); ++i)
-        point_values.row(dofs[i]) = values.row(i);
+        cells[dofs[i]] = c;
     }
 
+    eval(x_g, cells, point_values);
```
Comment (garth-wells) [inline cpp/dolfinx/fem/Function.h:409]: If necessary, `Function::eval` interface for `cells` can be changed to `tcb::span`.

**Review by garth-wells — CHANGES_REQUESTED** [review]

**Comment by massimiliano-leoni** [comment]
Here my new timings after @mscroggs 's latest update:
```2021-01-26 13:18:06.823 (  32.931s) [main            ]           XDMFFile.cpp:50    INFO| Opened HDF5 file with id "72057594037927938"
2021-01-26 13:18:06.823 (  32.931s) [main            ]          xdmf_mesh.cpp:197   INFO| Adding mesh to node "/Xdmf/Domain"
2021-01-26 13:18:06.823 (  32.932s) [main            ]          xdmf_mesh.cpp:25    INFO| Adding topology data to node "/Xdmf/Domain/Grid"
2021-01-26 13:18:06.900 (  33.008s) [main            ]          xdmf_mesh.cpp:146   INFO| Adding geometry data to node "/Xdmf/Domain/Grid"
2021-01-26 13:18:07.624 (  33.733s) [main            ]           XDMFFile.cpp:50    INFO| Opened HDF5 file with id "72057594037927939"
2021-01-26 13:18:07.624 (  33.733s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-26 13:18:07.898 (  34.007s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-26 13:18:08.162 (  34.271s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-26 13:18:08.426 (  34.535s) [main            ]      xdmf_function.cpp:84    INFO| Adding function to node "/Xdmf/Domain/Grid/Grid"
2021-01-26 13:18:08.685 (  34.794s) [main            ] NavierStokesSolver.cpp:272   INFO| Solving momentum
```
Less than 2 seconds in total, which is much faster than ever before.

**Comment by jorgensd** [comment]
Summarizing: 
The following MWE:
```
import dolfinx
import dolfinx.io
import dolfinx.common
from mpi4py import MPI
N = 40
mesh = dolfinx.UnitCubeMesh(MPI.COMM_WORLD, N, N, N)
V = dolfinx.VectorFunctionSpace(mesh, ("CG", 1))
u = dolfinx.Function(V)
with dolfinx.io.XDMFFile(MPI.COMM_WORLD, "mesh.xdmf", "w") as xdmf:
    with dolfinx.common.Timer("Write mesh"):
        xdmf.write_mesh(mesh)
    with dolfinx.common.Timer("Write function"):
        xdmf.write_function(u)
dolfinx.common.list_timings(MPI.COMM_WORLD, [dolfinx.common.TimingType.wall])
```
which pre `basix` had an average of 0.67 seconds in runtime (using 4 mpi ranks).
After the proposed restructuring of `compute_point_values` (such that basix only tabulate once per cell in `compute_reference_geometry`, and we avoid allocating an empty array for permutation data in `Function::eval`, the resulting timings with 4 ranks is: `0.31`.
Note that we introduce an interpolation of a function prior to calling `xdmf.write_function` (and in turn `Function.compute_point_values`), we obtain an even more significant speedup: (4 mpi ranks of the following program yields: `0.057177`).
```
import dolfinx
import dolfinx.io
import dolfinx.common
from mpi4py import MPI
import numpy as np
N = 40
mesh = dolfinx.UnitCubeMesh(MPI.COMM_WORLD, N, N, N)
V = dolfinx.VectorFunctionSpace(mesh, ("CG", 1))
u = dolfinx.Function(V)


def f(x):
    vals = np.zeros((3, x.shape[1]))
    vals[0] = x[0]
    vals[1] = x[0] + x[1]
    vals[2] = np.sin(np.pi * x[2])
    return vals


u.interpolate(f)
with dolfinx.io.XDMFFile(MPI.COMM_WORLD, "mesh.xdmf", "w") as xdmf:
    with dolfinx.common.Timer("Write mesh"):
        xdmf.write_mesh(mesh)
    with dolfinx.common.Timer("Write function"):
        xdmf.write_function(u)
timing_mesh = dolfinx.common.timing("Write mesh")[1]
timing_function = dolfinx.common.timing("Write function")[1]
dolfinx.common.list_timings(MPI.COMM_WORLD, [dolfinx.common.TimingType.wall])
```

**Review by garth-wells — APPROVED** [review]

**Pull request merged**